org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java

```diff
@@ -2277,23 +2277,11 @@ private String getAnnotationType(IJavaProject javaProject, List<String> annotati
 								}
 							}
 						}
+						return findTypeInProject(javaProject, annotationType, classpathStorage);
```

Comment:
Sorry for replying late, in this way, will the annotations in the configuration will be skipped except the first one? Since we will always return a value here.

Reply:
You are right. A potential patch:
```
diff --git a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java
index c0230e28..667fff24 100644
--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java
@@ -2277,11 +2277,17 @@ public class Preferences {
                                                                }
                                                        }
                                                }
-                                               return findTypeInProject(javaProject, annotationType, classpathStorage);
+                                               String aType = findTypeInProject(javaProject, annotationType, classpathStorage);
+                                               if (aType != null) {
+                                                       return aType;
+                                               }
                                        } else {
                                                // for unknown types, try to find type in the project
                                                try {
-                                                       return findTypeInProject(javaProject, annotationType, classpathStorage);
+                                                       String aType = findTypeInProject(javaProject, annotationType, classpathStorage);
+                                                       if (aType != null) {
+                                                               return aType;
+                                                       }
                                                } catch (JavaModelException e) {
                                                        continue;
                                                }
```

Reply:
The change looks good, with this patch, each null type in the configurations but not found in the classpath will trigger method `IJavaProject.findType()`, which *might* cause performance issues. WDYT? 

I noticed the discussion https://github.com/eclipse/eclipse.jdt.ls/pull/2486#discussion_r1113472506 and I'm also trying to find a way to improve a performance here. Maybe letting users specify a exact null type instead of null type array would be helpful?

Reply:
@CsCherrYY Could you, please, review https://github.com/eclipse/eclipse.jdt.ls/pull/2493 ?

>  Maybe letting users specify a exact null type instead of null type array would be helpful?

`IJavaProject.findType()` is only slow when called the first time.